Improved YNAB payee name for Hapoalim transfers

Nitaym opened the pull request:
Hey,

I ran into the issue where Hapoalim transfers would show up as "העברה מהבנק" or "העב' לאחר-נייד" where the payee name exists in the transaction memo. 
I'm not sure if that's the prettiest way to finding the payee name, but it does work... 

Let me know what you think

baruchiro commented:
Interesting idea. @brafdlog What do you think?

If we continue with this, I think we should consider it as a "small component", a new step of parsing description.

brafdlog commented:
Thanks @Nitaym 
I agree with @baruchiro that this could grow to contain more importer specific logic.
In any case, we don't want vendor specific customization in the main function.
Please extract this logic to a getPayeeName function and we can grow that logic as needed in the future

Nitaym commented:
Good idea! Did as suggested.  (Also fixed the linting issues)

I've pushed the changes, but will only be able to test it tonight hopefully. Moved it to draft in the meanwhile (unless you guys are able to test earlier)

baruchiro commented:
Can you write unit tests on it?

Nitaym commented:
Probably - Where can I find the existing ones?

Nitaym commented:
Hey @baruchiro, I've added tests. Please review

baruchiro commented on src/backend/export/outputVendors/ynab/ynab.test.ts:
```diff
@@ -55,4 +57,60 @@ describe('ynab', () => {
+      status: TransactionStatuses[TransactionStatuses.Completed],
+    }
+   
+    test('Verify getPayeeName extracts the correct payeeName', async () => {
+      transactionSample.description = 'מסטרקרד';
+      transactionSample.memo = '';
+      expect(ynab.getPayeeName(transactionSample)).toBe(transactionSample.description);
+      
+      transactionSample.description = ' משכורת';
+      transactionSample.memo = 'המבצע: לאומי';
+      expect(ynab.getPayeeName(transactionSample)).toBe(transactionSample.description);
+      
+      transactionSample.description = 'קצבת ילדים';
+      transactionSample.memo = '';
+      expect(ynab.getPayeeName(transactionSample)).toBe(transactionSample.description);
+  
+      transactionSample.description = 'בזק-הוראת קבע';
+      transactionSample.memo = 'בזק - חיובי טלפון'; 
+      expect(ynab.getPayeeName(transactionSample)).toBe(transactionSample.description);
+    });
```
```suggestion
    test.each([
      ['מסטרקרד', ''],
      [' משכורת', 'המבצע: לאומי'],
      ['קצבת ילדים', ''],
      ['בזק-הוראת קבע', 'בזק - חיובי טלפון']
    ])('Verify getPayeeName extracts the correct payeeName', async (description, memo) => {
      transactionSample.description = description;
      transactionSample.memo = memo;
      expect(ynab.getPayeeName(transactionSample)).toBe(transactionSample.description);
    });
```

baruchiro commented on src/backend/export/outputVendors/ynab/ynab.test.ts:
```diff
@@ -55,4 +57,60 @@ describe('ynab', () => {
+      transactionSample.memo = 'בזק - חיובי טלפון'; 
+      expect(ynab.getPayeeName(transactionSample)).toBe(transactionSample.description);
+    });
+    test('Verify hapoalim transfers capture the correct payee name', async () => {
+      transactionSample.description = 'הוראת קבע';
+      transactionSample.memo = 'לטובת: צהרון. עבור: צהרון גנים - ילד 00-000-0000000';
+      expect(ynab.getPayeeName(transactionSample)).toBe('צהרון');
+      
+      transactionSample.description = "העב' לאחר-נייד";
+      transactionSample.memo = 'לטובת: איש כלשהו. עבור: סוף חשבון';
+      expect(ynab.getPayeeName(transactionSample)).toBe('איש כלשהו');
+  
+      transactionSample.description = "העברה לאחר";
+      transactionSample.memo = 'לטובת: פנסיה לדוגמא. עבור: סיבה מסויימת';
+      expect(ynab.getPayeeName(transactionSample)).toBe('פנסיה לדוגמא');
+  
+      transactionSample.description = 'העברה מהבנק';
+      transactionSample.memo = 'לטובת: אישה כלשהי. עבור: משכורת אוגוסט';
+      expect(ynab.getPayeeName(transactionSample)).toBe('אישה כלשהי');
+      
+      // Weird bank accounts names
+      transactionSample.description = 'הוראת קבע';
+      transactionSample.memo = 'לטובת: di. עבור: גן 092-.';  
+      expect(ynab.getPayeeName(transactionSample)).toBe('di');
+    });
```
```suggestion
    test.each([
      ['הוראת קבע', 'לטובת: צהרון. עבור: צהרון גנים - ילד 00-000-0000000', 'צהרון'],
      ["העב' לאחר-נייד", 'לטובת: איש כלשהו. עבור: סוף חשבון', 'איש כלשהו'],
      ['העברה לאחר', 'לטובת: פנסיה לדוגמא. עבור: סיבה מסויימת', 'פנסיה לדוגמא'],
      ['העברה מהבנק', 'לטובת: אישה כלשהי. עבור: משכורת אוגוסט', 'אישה כלשהי'],
      // Weird bank accounts names
      ['הוראת קבע', 'לטובת: di. עבור: גן 092-.', 'di'],
    ])('Verify hapoalim transfers capture the correct payee name', async (description, memo, expected) => {
      transactionSample.description = description;
      transactionSample.memo = memo;
      expect(ynab.getPayeeName(transactionSample)).toBe(expected);
    });
```

baruchiro commented on src/backend/export/outputVendors/ynab/ynab.ts:
```diff
@@ -75,16 +75,28 @@ function getTransactions(startDate: Date): Promise<ynab.TransactionsResponse> {
   return ynabAPI!.transactions.getTransactions(ynabConfig!.options.budgetId, moment(startDate).format(YNAB_DATE_FORMAT));
 }
 
+export function getPayeeName(transaction: EnrichedTransaction, payeeNameMaxLength: number = 50) {
+  let payeeName = transaction.description.substring(0, payeeNameMaxLength);
+
+  // Specific case for Bank Hapoalim - Where we can extract the payee name from the memo. Note the "." is added 
+  // from the israeli-bank-scrapers project, and not from the bank itself
+  if (transaction.memo?.startsWith('לטובת') && transaction.memo.includes('עבור')) {
+    payeeName = transaction.memo.slice(7, transaction.memo.indexOf('עבור') - 2);
+  }
+
+  return payeeName;
+}
```
```suggestion
export function getPayeeName(transaction: EnrichedTransaction, payeeNameMaxLength: number = 50) {
  // Specific case for Bank Hapoalim - Where we can extract the payee name from the memo. Note the "." is added
  // from the israeli-bank-scrapers project, and not from the bank itself
  if (transaction.memo?.startsWith('לטובת') && transaction.memo.includes('עבור')) {
    return transaction.memo.slice(7, transaction.memo.indexOf('עבור') - 2);
  }

  return transaction.description.substring(0, payeeNameMaxLength);
}
```

baruchiro reviewed (changes requested):
Good work!

A few suggestions.

baruchiro commented:
Also, you need to fix lint issues.

Nitaym commented:
Thanks for the suggestions!

Nitaym commented:
Changes done

brafdlog commented on src/backend/export/outputVendors/ynab/ynab.ts:
```diff
@@ -75,16 +75,25 @@ function getTransactions(startDate: Date): Promise<ynab.TransactionsResponse> {
     amount,
     // "payee_id": "string",
-    payee_name: originalTransaction.description.substring(0, payeeNameMaxLength),
+    payee_name: getPayeeName(originalTransaction, ynabConfig!.options.maxPayeeNameLength || 50),
```
Isn't the || 50 redundant now that you made 50 the default value of the parameter?

Nitaym replied:
I'm not a js dev so bear with me,

if `ynabConfig!.options.maxPayeeNameLength` is undefined, won't `getPayeeName` get the value `undefined` or `null` or whatever value `maxPayeeNameLength` has?

brafdlog replied:
In general yes, but you are calling the getPayeeName function and passing this as the second parameter.
For this parameter you defined a default value of 50
`export function getPayeeName(transaction: EnrichedTransaction, payeeNameMaxLength: number = 50) {`

So if you don't pass a parameter or pass undefined (which should be the value of this if no one assigned any value to id) it will get the default value

See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters

baruchiro reviewed (approved):
Oh sorry, @Nitaym @brafdlog I forgot I did it 😄 

![image](https://user-images.githubusercontent.com/17686879/173999404-8065804d-fa08-4a77-b7e8-29c13995aafa.png)

@Nitaym What is your next contribution? 😉

Nitaym commented:
Thanks!

I'm a user - Wherever I see a need - I add a fix :)
